sdk/python/lib/pulumi/runtime/stack.py

```diff
@@ -69,8 +66,11 @@
         # and each task needs to get scheduled in order to acknowledge the cancel and exit.
         await asyncio.sleep(0)
 
-        # Once we get scheduled again, all tasks have exited and we're good to go.
-        log.debug("run_in_stack completed")
+    # Once we get scheduled again, all tasks have exited and we're good to go.
+    log.debug("run_in_stack completed")
+
+    if RPC_MANAGER.unhandled_exception is not None:
+        raise RPC_MANAGER.unhandled_exception.with_traceback(RPC_MANAGER.exception_traceback)
```

Comment:
The assumption here seems to be that this only needs to be checked if the program synchronously completed successfully.  Is that true?  I suppose the only other way to get here would be another exception already being propagated, and so in that case we don't need/want to raise the unhandled exception. Though if it is possible for that to happen, that seems like another potential source of confusion.

Reply:
As it happens, we already log on any exception that occurs in the context of `do_rpc`.